src/Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv/Internal/LibuvOutputConsumer.cs

```diff
@@ -47,14 +47,22 @@ public class LibuvOutputConsumer
                     if (!buffer.IsEmpty)
                     {
                         var writeReq = _writeReqPool.Allocate();
-                        var writeResult = await writeReq.WriteAsync(_socket, buffer);
-                        _writeReqPool.Return(writeReq);
 
-                        LogWriteInfo(writeResult.Status, writeResult.Error);
+                        try
+                        {
+                            var writeResult = await writeReq.WriteAsync(_socket, buffer);
+
+                            LogWriteInfo(writeResult.Status, writeResult.Error);
 
-                        if (writeResult.Error != null)
+                            if (writeResult.Error != null)
+                            {
+                                throw writeResult.Error;
+                            }
+                        }
+                        finally
                         {
-                            throw writeResult.Error;
+                            // Make sure we return the writeReq to the pool
+                            _writeReqPool.Return(writeReq);
```

Comment:
Why add a nested try/finally? Can WriteAsync throw? This code makes it look like it could. What was wrong with the old way of returning this?